extension/opampextension/opamp_agent.go

```diff
@@ -0,0 +1,244 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package opampextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/opampextension"
+
+import (
+	"context"
+	"crypto/tls"
+	"net/http"
+	"os"
+	"runtime"
+
+	"github.com/oklog/ulid/v2"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	semconv "go.opentelemetry.io/collector/semconv/v1.18.0"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opamp-go/client"
+	"github.com/open-telemetry/opamp-go/client/types"
+	"github.com/open-telemetry/opamp-go/protobufs"
+)
+
+// TODO: Replace with https://github.com/open-telemetry/opentelemetry-collector/issues/6596
+const localConfig = `
+exporters:
+  otlp:
+    endpoint: localhost:1111
+receivers:
+  otlp:
+    protocols:
+      grpc: {}
+      http: {}
+service:
+  pipelines:
+    traces:
+      receivers: [otlp]
+      processors: []
+      exporters: [otlp]
+`
+
+type opampAgent struct {
+	cfg    *Config
+	logger *zap.Logger
+
+	agentType    string
+	agentVersion string
+
+	instanceId ulid.ULID
+
+	effectiveConfig string
+
+	agentDescription *protobufs.AgentDescription
+
+	opampClient client.OpAMPClient
+
+	remoteConfigStatus *protobufs.RemoteConfigStatus
+}
+
+func (o *opampAgent) Start(_ context.Context, _ component.Host) error {
+	o.opampClient = client.NewWebSocket(o.logger.Sugar())
+
+	header := http.Header{}
+	for k, v := range o.cfg.Headers {
+		header.Set(k, string(v))
+	}
+
+	tls := &tls.Config{
+		ServerName:         o.cfg.TLSSetting.ServerName,
+		InsecureSkipVerify: o.cfg.TLSSetting.InsecureSkipVerify,
+	}
+
+	settings := types.StartSettings{
+		Header:         header,
+		TLSConfig:      tls,
+		OpAMPServerURL: o.cfg.Endpoint,
+		InstanceUid:    o.instanceId.String(),
+		Callbacks: types.CallbacksStruct{
+			OnConnectFunc: func() {
+				o.logger.Debug("Connected to the OpAMP server")
+			},
+			OnConnectFailedFunc: func(err error) {
+				o.logger.Error("Failed to connect to the OpAMP server", zap.Error(err))
+			},
+			OnErrorFunc: func(err *protobufs.ServerErrorResponse) {
+				o.logger.Error("OpAMP server returned an error response", zap.String("message", err.ErrorMessage))
+			},
+			GetEffectiveConfigFunc: func(ctx context.Context) (*protobufs.EffectiveConfig, error) {
+				return o.composeEffectiveConfig(), nil
+			},
+			OnMessageFunc: o.onMessage,
+		},
+		// Include ReportsEffectiveConfig once the extension has access to the
+		// collector's effective configuration.
+		Capabilities: protobufs.AgentCapabilities_AgentCapabilities_ReportsStatus,
+	}
+
+	if err := o.createAgentDescription(); err != nil {
+		return err
+	}
+
+	if err := o.opampClient.SetAgentDescription(o.agentDescription); err != nil {
+		return err
+	}
+
+	o.logger.Debug("Starting OpAMP client...")
+
+	if err := o.opampClient.Start(context.Background(), settings); err != nil {
```

Comment:
An error here will bring down the collector which should only happen in unrecoverable circumstances. I see that the current implementation of `client.Start` only errs if it cannot parse the URL, so this appears to be fine. However, I'm curious if this is expected by the protocol or subject to change.

Reply:
The `Start()` contract currently says:
```
	// Start may immediately return an error if the settings are incorrect (e.g. the
	// serverURL is not a valid URL).
	// Start does not wait until the connection to the Server is established and will
	// likely return before the connection attempts are even made.
```
Do we need to say more to make it clear that it won't error for transient errors?

Reply:
`Start()` may check for more conditions than `Validate()` does so ignoring it may be wrong.

Reply:
I must be missing a nuance there but I'll resolve the comment based on your assurance that it `client.Start` does not fail on transient errors.

Reply:
>I must be missing a nuance there

As an example: `Start()` will fail if `SetAgentDescription()` is not already called. `Config.Validate` obviously knows nothing about and can't do that validation.